xds/internal/balancer/ringhash/ringhash.go

```diff
@@ -121,3 +153,276 @@ func (sc *subConn) queueConnect() {
 	// after backoff in TransientFailure), it will Connect().
 	sc.connectQueued = true
 }
+
+type ringhashBalancer struct {
+	cc     balancer.ClientConn
+	logger *grpclog.PrefixLogger
+
+	config *LBConfig
+
+	subConns map[resolver.Address]*subConn // `attributes` is stripped from the keys of this map (the addresses)
+	scStates map[balancer.SubConn]*subConn
+
+	// ring is always in sync with subConns. When subConns change, a new ring is
+	// generated. Note that address weights updates (they are keys in the
+	// subConns map) also regenerates the ring.
+	ring    *ring
+	picker  balancer.Picker
+	csEvltr *connectivityStateEvaluator
+	state   connectivity.State
+
+	resolverErr error // the last error reported by the resolver; cleared on successful resolution
+	connErr     error // the last connection error; cleared upon leaving TransientFailure
+}
+
+// updateAddresses creates new SubConns and removes SubConns, based on the
+// address update.
+//
+// The return value is whether the new address list is different from the
+// previous. True if
+// - an addresses was added
+// - an addresses was removed
+// - an address's weight was updated
+//
+// Note that this function doesn't trigger SubConn connecting, so all the new
+// SubConn states are Idle.
+func (b *ringhashBalancer) updateAddresses(addrs []resolver.Address) bool {
+	var addrsUpdated bool
+	// addrsSet is the set converted from addrs, it's used for quick lookup of
+	// an address.
+	//
+	// Addresses in this map all have attributes stripped, but metadata set to
+	// the weight. So that weight change can be detected.
+	//
+	// TODO: this won't be necessary if there are ways to compare address
+	// attributes.
```

Comment:
Should we mention this change is needed in the open issue?  Or file a new issue?

Reply:
Made #4768